spmd/tensor/parallel/api.py

```diff
@@ -1,122 +1,194 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
 import torch
 import torch.nn as nn
-from typing import Sequence, Tuple
+from typing import Union, Dict
 from spmd.tensor import (
     distribute_module,
     distribute_tensor,
-    DTensor,
     Shard,
     Replicate,
     DeviceMesh,
-    Placement,
 )
 from spmd.tensor.parallel import TensorParallelMultiheadAttention
-from spmd.tensor.parallel.style import ParallelStyle, PairwiseParallel
+from spmd.tensor.parallel.style import PairwiseParallel, ParallelStyle
 from spmd.tensor.parallel.utils import _create_1d_device_mesh
 
 
-def replicate_input(
-    inputs: Sequence[torch.Tensor], device_mesh: DeviceMesh
-) -> Tuple[DTensor, ...]:
-    replicate = [Replicate()] * device_mesh.ndim
-    return tuple(
-        DTensor.from_local(tensor, device_mesh, replicate) for tensor in inputs
-    )
+__all__ = [
+    "parallelize_module",
+]
 
 
-def replicate_output(output: DTensor, device_mesh: DeviceMesh) -> torch.Tensor:
-    if isinstance(output, DTensor):
-        replicate = [Replicate()] * output.device_mesh.ndim
-        # TODO: can the output be left incontiguous?
-        return (
-            output.redistribute(output.device_mesh, replicate)
-            .to_local()
-            .contiguous()
-        )
+def parallelize_module(  # type: ignore[return]
+    module: nn.Module,
+    device_mesh: DeviceMesh,
+    parallelize_plan: Union[ParallelStyle, Dict[str, ParallelStyle]],
+    tp_mesh_dim: int = 0,
+) -> nn.Module:
+    """
+    The major API for tensor parallelism in PyTorch. We parallelize module
```

Comment:
This is the only API we plan to expose to the user right? Let's just call it "The API to apply Tensor Parallelism"?

Reply:
Yep, sure, let me change it.